samples/matter/common/src/event_triggers/default_event_triggers.cpp

```diff
@@ -20,19 +21,75 @@ using namespace chip;
 
 namespace
 {
-k_timer sFactoryResetDelayTimer;
-
-void FactoryResetDelayTimerTimeoutCallback(k_timer *timer)
+/* Timer to delay some operations to allow send back ACK to Matter Controller. */
+k_timer sDelayTimer;
+bool sDelayedActionInProgress = false;
+
+/* Some operations may interrupt communication with the Matter controller so we need to postpone the action because it
+ * must receive the acknowledgement from the device.  */
+enum class DelayedAction { FactoryReset, BlockMainThread };
+struct DelayedCtx {
+	DelayedAction action;
+	TestEventTrigger::TriggerValue value;
+};
+
+void DelayTimerCallback(k_timer *timer)
 {
-	Server::GetInstance().ScheduleFactoryReset();
+	/* Schedule an action in the Main thread to allow releasing Delayed context memory (it cannot be done within the
+	 * Timer IRQ)*/
+	Nrf::PostTask([timer] {
+		Platform::UniquePtr<DelayedCtx> ctx(reinterpret_cast<DelayedCtx *>(k_timer_user_data_get(timer)));
+		if (!ctx) {
+			return;
+		}
+
+		TestEventTrigger::TriggerValue value = ctx->value;
+
+		switch (ctx->action) {
+		case DelayedAction::FactoryReset:
+			Server::GetInstance().ScheduleFactoryReset();
+			break;
+		case DelayedAction::BlockMainThread:
+			Nrf::PostTask([value] {
+				int64_t startTime = k_uptime_get();
+
+				LOG_INF("Blocking Main thread for %d ms", value);
+
+				while (startTime + value > k_uptime_get()) {
+					;
+				}
+
+				LOG_INF("Blocking Main thread finished");
+			});
+			break;
+		default:
+			break;
+		}
+	});
+
+	/* Assign that the delayed action has been finished */
+	sDelayedActionInProgress = false;
```

Comment:
You still use the timer user data after setting this (as you call `k_timer_user_data_get` from the application thread), which means the user data may be overwritten before the previous is consumed.

Reply:
Is better now? :)

Reply:
yes, thanks :)